core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -659,6 +666,56 @@ class LogManager(logDirs: Seq[File],
       Option(currentLogs.get(topicPartition))
   }
 
+  /**
+   * Method to indicate that logs are getting initialized for the partition passed in as argument.
+   * This method should always be followed by [[kafka.log.LogManager#finishedInitializingLog]] to indicate that log
+   * initialization is done.
+   */
+  def initializingLog(topicPartition: TopicPartition): Unit = configUpdateLock synchronized {
+    partitionsInitializing(topicPartition) = false
+  }
+
+  /**
+   * Mark the partition configuration for partitions that are getting initialized for topic
+   * as dirty. That will result in reloading of configuration once initialization is done.
+   */
+  def topicConfigUpdated(topic: String): Unit = configUpdateLock synchronized {
+    partitionsInitializing.collect {
+      case (topicPartition, _) if topicPartition.topic() == topic => partitionsInitializing(topicPartition) = true
+    }
+  }
+
+  /**
+   * Mark all in progress partitions having dirty configuration if broker configuration is updated.
+   */
+  def brokerConfigUpdated(): Unit = configUpdateLock synchronized {
+    partitionsInitializing.foreach {
+      case (topicPartition, _) => partitionsInitializing(topicPartition) = true
+    }
+  }
+
+  /**
+   * Method to indicate that the log initialization for the partition passed in as argument is
+   * finished. This method should follow a call to [[kafka.log.LogManager#initializingLog]]
+   */
+  def finishedInitializingLog(topicPartition: TopicPartition,
+                              maybeLog: Option[Log],
+                              config: => LogConfig): Unit = {
+    /*
+     * Keep on reloading config as long as it keeps changing. A topic or broker config change
+     * event (in ConfigHandler::processConfigChanges or DynamicLogConfig::reconfigure) can
+     * arrive while we are loading config. If that happens reload the configuration.
+     */
+    while (partitionsInitializing(topicPartition)) {
```

Comment:
I think I understand the case we're trying to protect, but possibly it borders on overkill because I'm not sure how we can guarantee a correct ordering of updates in all cases. For example, in `TopicConfigHandler. updateLogConfig`, we have a similar race condition where we fetch the config, then update it.

Reply:
Done. Replaced `while` with `if`.